### ISSUE TITLE
multiline HTML comments are improperly encoded

### ISSUE THREAD

**Issue by zeke** [markdown-it/markdown-it#51]
```js
console.log(
  require('markdown-it')({html: true}).render(
    "<!-- one-liner -->\n\n<!-- two-liner\n\n-->"
  )
)
```

The closing tag on the two-line comment is getting encoded:

```
<!-- one-liner -->
<!-- two-liner
<p>--&gt;</p>
```

Is this expected behavior or a bug?

// cc @rwaldron who opened https://github.com/npm/newww/issues/631

**Comment by puzrin** [comment]
Yes, it's expected. In commonmark html blocks (not inlines in paragraph) are inserted as raw lines, and empty line means the end.

Some upcoming changes are discussed here http://talk.commonmark.org/t/raw-html-blocks-proposals-comments-wanted/983. But in general, comments will not pass between blocks.

You can try to cheat with replacing `html_block` rule to something custom.

**Comment by puzrin** [comment]
I've looked into src https://raw.githubusercontent.com/rwaldron/johnny-five/master/README.md

Well, if this case is often, you can add new block rule specially for such comments.

**Comment by cliss** [comment]
This is something I run into often. Other parsers such as [marked](https://github.com/chjj/marked) treat them as I would expect &mdash; everything between an opening and closing tag is treated as HTML. This is particularly relevant to me in the case of `<script />` tags.

I'd be happy to write a plugin (or, as @puzrin mentioned, a new block rule), but honestly I feel like I'm a bit over my head here. If anyone could offer a bit more specific guidance/direction, I'd be much obliged.

**Comment by puzrin** [comment]
Generic guide for developpers is here https://github.com/markdown-it/markdown-it/tree/master/docs

This rule needs spec, because edge cases possible. For example, rule starts at the begining of line, but finishes in the middle of string, and has some tailing text, and so on. Probably it worth to discuss it [here](http://talk.commonmark.org/t/raw-html-blocks-proposals-comments-wanted/983/30?u=vitaly) first. 

I see no technical problem to do it from `fensed block` rule, but until CommonMark spec updated, prefer to have it in plugin, not incore. Reason - avoid creating "one more standard".

Current implementation has divergence only for tables & strikes. Those are "something like github does, but nobody guarantee anything".

**Comment by puzrin** [comment]
Related ticket in commonmark tracker https://github.com/jgm/CommonMark/issues/177

As jgm replied at forum, "html block" improvments have high priority.

I'd prefer to wait spec update. Then improvments will be done in this package very fast. But if anyone can't wait at all - write a plugin.

**Comment by zeke** [comment]
Thanks for investigating, @puzrin!

**Comment by zeke** [comment]
> You can try to cheat with replacing html_block rule to something custom.

I was initially happy to wait for the [spec change](https://github.com/jgm/CommonMark/issues/177) to take effect, but we continue to receive reports of npm READMEs with this multiline comment problem. @puzrin can you provide some insight on how to properly override `html_block`?

- https://github.com/npm/newww/issues/866
- https://github.com/npm/newww/issues/631
- https://github.com/npm/newww/issues/622

**Comment by puzrin** [comment]
-  https://github.com/npm/newww/issues/866 - can't reproduce. I've copied readme content to markdown-it demo and don't see any problems.

As far as i see, the most of problems happens due bad/incompatible block comments closer handling. I would hack it with new additional block rule, specially for comments. Something like simplified [fenced block](https://github.com/markdown-it/markdown-it/blob/master/lib/rules_block/fence.js).

- it should check `<!--` at start of line (compare by charcodes for speed)
- if found - seek down until line with `-->` found
- on rule add it should have [flags](https://github.com/markdown-it/markdown-it/blob/master/lib/parser_block.js#L16), that in can terminate paragraph, list and blockquote (`alt` option). And priority should be before existing `html_block`.

It still unclear what to to ih this case:

```
--> text after closing block comment
```

But you had no such issues, and i think tail can be dropped silently, untill someone report new bug. Or you can cheat with string start/shift markers to make tail work as separate line on next pass.

**Comment by puzrin** [comment]
Also, it worth to show better error text instead of "no readme". Something like "please, reformat html blocks and comments in you readme to follow [CommonMark] spec - no emply line in such blocks and one empty line at the end".

**Comment by puzrin** [comment]
Good news, spec progress moved forward: http://talk.commonmark.org/t/raw-html-blocks-proposals-comments-wanted/983/52. I expect this problem to be solved in 1-2 month, including iterations for bugfixes and spec clarifications.

Join to discussion, if you have suggestions for the spec.

**Comment by puzrin** [comment]
Fixed in 4.4.0. See CM 0.21 spec change http://spec.commonmark.org/
